rtrlib/aspa/aspa_array/aspa_array.c

```diff
@@ -23,140 +23,222 @@ int aspa_array_create(struct aspa_array **vector_pointer)
-	vector->capacity = default_initial_size;
-	vector->size = 0;
-	vector->data = data_field;
+	array->capacity = default_initial_size;
+	array->size = 0;
+	array->data = data_field;
 
 	// returning the vector
-	*vector_pointer = vector;
+	*array_ptr = array;
 
-	return 0;
+	return ASPA_SUCCESS;
 }
 
-int aspa_array_free(struct aspa_array *vector)
+enum aspa_status aspa_array_free(struct aspa_array *array, bool free_provider_sets)
 {
 	// if the vector is null just return
-	if (vector == NULL) {
-		return -1;
+	if (!array) {
+		return ASPA_ERROR;
 	}
 
-	if (vector->data != NULL) {
+	if (array->data) {
 		// freeing the data
-		lrtr_free(vector->data);
+		lrtr_free(array->data);
+
+		if (free_provider_sets) {
+			for (size_t i = 0; i < array->size; i++) {
+				if (array->data[i].provider_asns) {
+					lrtr_free(array->data[i].provider_asns);
+					array->data[i].provider_asns = NULL;
+				}
+			}
+		}
 	}
 
-	// freeing the object itself
-	lrtr_free(vector);
+	// freeing the array itself
+	lrtr_free(array);
 
-	return 0;
+	return ASPA_SUCCESS;
 }
 
-int aspa_array_reallocate(struct aspa_array *vector)
+enum aspa_status aspa_array_reallocate(struct aspa_array *array)
 {
 	// the factor by how much the capacity will increase: new_capacity = old_capacity * SIZE_INCREASE_EXPONENTIAL
 	const size_t SIZE_INCREASE_EXPONENTIAL = 2;
 
 	// allocation the new chunk of memory
 	struct aspa_record *new_data_field = (struct aspa_record *)lrtr_malloc(
-		sizeof(struct aspa_record) * vector->capacity * SIZE_INCREASE_EXPONENTIAL);
+		sizeof(struct aspa_record) * array->capacity * SIZE_INCREASE_EXPONENTIAL);
 
 	// malloc failed so returning an error
 	if (new_data_field == NULL) {
-		return -1;
+		return ASPA_ERROR;
 	}
 
 	// copying the data from the old location to the new one
-	memcpy(new_data_field, vector->data, vector->capacity * sizeof(struct aspa_record));
+	memcpy(new_data_field, array->data, array->capacity * sizeof(struct aspa_record));
 
 	// deleting the old vector
-	lrtr_free(vector->data);
+	lrtr_free(array->data);
 
 	// assigning the new array to the vector and incrementing the capacity
-	vector->data = new_data_field;
-	vector->capacity *= SIZE_INCREASE_EXPONENTIAL;
+	array->data = new_data_field;
+	array->capacity *= SIZE_INCREASE_EXPONENTIAL;
 
-	return 0;
+	return ASPA_SUCCESS;
 }
 
-enum aspa_rtvals aspa_array_private_insert(struct aspa_array *vector, struct aspa_record record, bool overwrite) {
-	// TODO: Handle replacements
-	// iterator running from the back of the array to the front
-	size_t j = vector->size;
-
-	// we are moving from the back of the array to the front while looking
-	// for the place where to insert our new record
-	while (j > 0 && vector->data[j - 1].customer_asn > record.customer_asn) {
-		// moving data to the right
-		vector->data[j] = vector->data[j - 1];
-		j -= 1;
-	}
+enum aspa_status aspa_array_set_record(struct aspa_array *array, size_t index, struct aspa_record *record)
```

Comment:
no bounds check ? / garbage collection ?

Reply:
oh right I forgot

Reply:
I know why.. `aspa_array_set_record` isn't called anywhere and can be removed